server/src/main/java/com/tencent/rss/server/buffer/ShuffleBuffer.java

```diff
@@ -0,0 +1,374 @@
+  // the strategy ensure that shuffle is in memory or storage
+  private List<ShufflePartitionedBlock> blocks;
+  private Map<Long, List<ShufflePartitionedBlock>> inFlushBlockMap;
+  private static final long INVALID_EVENT_ID = -1;
+
+  public ShuffleBuffer(long capacity) {
+    this.capacity = capacity;
+    this.size = 0;
+    this.blocks = new LinkedList<>();
+    this.inFlushBlockMap = Maps.newConcurrentMap();
+  }
+
+  public int append(ShufflePartitionedData data) {
+    int mSize = 0;
+
+    synchronized (this) {
+      for (ShufflePartitionedBlock block : data.getBlockList()) {
+        blocks.add(block);
+        mSize += block.getSize();
+        size += mSize;
+      }
+    }
+
+    return mSize;
+  }
+
+  public synchronized ShuffleDataFlushEvent toFlushEvent(
+      String appId,
+      int shuffleId,
+      int startPartition,
+      int endPartition,
+      Map<String, Map<Integer, AtomicLong>> shuffleSizeMap,
+      Supplier<Boolean> isValid) {
+    if (blocks.isEmpty()) {
+      return null;
+    }
+    // buffer will be cleared, and new list must be created for async flush
+    List<ShufflePartitionedBlock> spBlocks = new LinkedList<>(blocks);
+    long eventId = ShuffleFlushManager.ATOMIC_EVENT_ID.getAndIncrement();
+    final ShuffleDataFlushEvent event = new ShuffleDataFlushEvent(
+        eventId,
+        appId,
+        shuffleId,
+        startPartition,
+        endPartition,
+        size,
+        spBlocks,
+        isValid,
+        this);
+    inFlushBlockMap.put(eventId, spBlocks);
+    blocks.clear();
+    updateShuffleSize(shuffleSizeMap, appId, shuffleId, size);
+    size = 0;
+    return event;
+  }
+
+  // update metadata of shuffleId -> size
+  private void updateShuffleSize(
+      Map<String, Map<Integer, AtomicLong>> shuffleSizeMap,
+      String appId,
+      Integer shuffleId,
+      long flushSize) {
+    Map<Integer, AtomicLong> shuffleIdToSize = shuffleSizeMap.get(appId);
+    if (shuffleIdToSize != null) {
+      AtomicLong atomicLong = shuffleIdToSize.get(shuffleId);
+      if (atomicLong != null) {
+        atomicLong.addAndGet(-flushSize);
+      }
+    }
+  }
+
+  public List<ShufflePartitionedBlock> getBlocks() {
+    return blocks;
+  }
+
+  public long getSize() {
+    return size;
+  }
+
+  public boolean isFull() {
+    return size > capacity;
+  }
+
+  public synchronized void clearInFlushBuffer(long eventId) {
+    inFlushBlockMap.remove(eventId);
+  }
+
+  @VisibleForTesting
+  public Map<Long, List<ShufflePartitionedBlock>> getInFlushBlockMap() {
+    return inFlushBlockMap;
+  }
+
+  // 1. generate buffer segments and other info: if blockId exist, start with which eventId
+  // 2. according to info from step 1, generate data
+  public synchronized ShuffleDataResult getShuffleData(
+      long lastBlockId, int readBufferSize) {
+    try {
+      List<BufferSegment> bufferSegments = Lists.newArrayList();
+      ActualReadInfo actualReadInfo = updateBufferSegments(
```

Comment:
BufferSegment should be a field of ActualReadInfo.

Reply:
ok

Reply:
Actually ActualReadInfo should have a field which type is List<ShufflePartitionedBlock>. This field will be updated when buffersegment is updated. And we needn't found any block, when we copy data.

Reply:
ActualReadInfo will be removed